CustomSelectControl: Add flag for larger default size

mirka opened the pull request:
Part of #39397

## What?

Adds an internal `__next36pxDefaultSize` feature flag prop to introduce the larger default size for `CustomSelectControl`.

## Why?

In preparation for standardizing the default sizes to 36px height.

## How?

Following the procedure outlined in #39397, conditional styles have been added to enlarge the height and paddings. The font size is also enlarged to be consistent with `SelectControl`.

The inconsistent chevron (#39400) will be addressed separately.

@jasmussen How do the paddings feel?

## Testing Instructions

1. `npm run storybook:dev` and see the stories for `CustomSelectControl`.
2. From the Controls tab, toggle the `__next36pxDefaultSize` prop.

## Screenshots

### `__next36pxDefaultSize` = false
<img src="https://user-images.githubusercontent.com/555336/157986345-3a7dee16-ff2b-45dc-b05e-fedbb9139ace.png" alt="Current default size" width="600">

### `__next36pxDefaultSize` = true

_Updated with 16px padding:_

<img width="600" alt="Larger default size" src="https://user-images.githubusercontent.com/555336/158231916-a3957ade-c5ad-4786-b5b3-81b1b49465b0.png">

mirka committed:
Migrate stories

mirka committed:
CustomSelectControl: Add flag for larger default size

mirka committed:
Add changelog entry

jasmussen commented:
Nice, thank you for the clear instructions. This is what I see in Storybook:
![select](https://user-images.githubusercontent.com/1204802/158124034-d1d07e78-77ec-4d10-9617-4c80f598be6e.gif)

Looks good, looks more coherent. 

Just for anyone else reviewing, we've had conversations about standardizing the heights across all inputs to be 36px, potentially with an option for 40px as well in the future. As we continue to refine those inputs and sizes, just unifying it all under a single 36px value is a good start, and will make future changes simpler.

In that light, do we need to include `36px` in the flag at all? I.e. could it simpluy be called `__nextDefaultSize`? While the current plan is to standardize all on 36px and have 40px as an option, we could go with `__nextDefaultSize` and `__nextLargeSize` instead, and it'd be a bit more agnostic to the values. 

On the padding, here's a screenshot of the resting state:

<img width="224" alt="resting state of dropdown" src="https://user-images.githubusercontent.com/1204802/158124518-f2af0a32-326a-4e42-b1c7-09155dc8dce2.png">

Here's one of the target designs:
<img width="315" alt="target design" src="https://user-images.githubusercontent.com/1204802/158124894-da8a56cc-a2b3-465d-bd9e-9feb201f3aa6.png">

Here's a direct comparison of the metrics:
<img width="273" alt="metrics" src="https://user-images.githubusercontent.com/1204802/158125014-caa4eda0-6ac4-473b-895b-c5f567e13eb5.png">

A few small tweaks needed here. The left and right padding should be 16px (`$grid-unit-20` in the SCSS). For the chevron icon, we might need to pull it to the right a bit to massage the pixels, but it becomes visible when you compare with a input with the custom unit selector:
<img width="182" alt="custom units" src="https://user-images.githubusercontent.com/1204802/158125357-c7dcca60-7170-45ad-9a37-5fbc6e53cf54.png">

<img width="271" alt="dropdown" src="https://user-images.githubusercontent.com/1204802/158125368-729a6410-e263-463b-8b08-19e0bd3827c6.png">

There's a new chevronDownSmall icon, which we can use to better match the design, shown on the right here:

<img width="92" alt="chevronicons" src="https://user-images.githubusercontent.com/1204802/158125554-6e8722b3-b49f-408a-adf9-caa4151a096a.png">

The color of the border (both for the dropdown and all input fields) should be #949494, or `$gray-600` in the SCSS system. That matches 3:1 UI contrast ratio.

You note an alignment issue with the chevron needing a separate fix, sounds good, I don't think we need to do all of the above feedback in a single PR, but I figured I'd write it down regardless! Hope that's useful, thanks again!

ciampo commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -141,8 +143,11 @@ export default function CustomSelectControl( {
+						'components-custom-select-control__button',
+						{ 'is-next-36px-default-size': __next36pxDefaultSize }
+					),
+					isSmall: ! __next36pxDefaultSize,
```
Love it when we rely on composability instead of style overrides

ciampo reviewed (commented):
Code changes LGTM, and thank you for refactoring the Story to use Controls!

I'll let you and @jasmussen discuss and refine the component's spacing!

> In that light, do we need to include 36px in the flag at all? I.e. could it simply be called `__nextDefaultSize`? While the current plan is to standardize all on 36px and have 40px as an option, we could go with `__nextDefaultSize` and `__nextLargeSize` instead, and it'd be a bit more agnostic to the values.

I don't have a strong opinion here, both options sound good to me,

mirka commented:
Thanks for the reviews!

> In that light, do we need to include `36px` in the flag at all? I.e. could it simply be called `__nextDefaultSize`?

Good question. It isn't pretty, but the reason I prefer including a specific pixel value here is twofold:

1. For anyone who is not familiar with our high-level component plans, I think that specificity is actually useful. Even for me, I get confused with all the size differences and the verbose naming decreases my cognitive load.
2. I want to avoid creating a situation where this flag ends up adding even more chaos to the library. The main thing I'm hedging against is when these size migrations take much longer than expected to complete (e.g. I'm hit by a bus, or priorities change). I don't want a vague `__nextDefaultSize` flag lying around when the next design refresh comes along 😂

> The left and right padding should be 16px (`$grid-unit-20` in the SCSS).

Do you have any concerns about this diverging from the `Button` paddings (12px) at 36px height? For example if a `Select` and a `Button` were stacked vertically, the paddings would be slightly off.

> The color of the border (both for the dropdown and all input fields) should be #949494, or `$gray-600` in the SCSS system.

Good to know. I've split out a separate issue (#39425) so we can change all of them at once.

jasmussen commented:
> Do you have any concerns about this diverging from the Button paddings (12px) at 36px height? For example if a Select and a Button were stacked vertically, the paddings would be slightly off.

Excellent eye for detail. Thank you.

It's my sense that we'd want to standardize on 16px padding left and right for all of these 36px items, including Buttons (barring any responsive changes we might make). That's perhaps the biggest benefit this effort has: unifying all these drifted values across a single grid.

I'd think it okay to not do those changes all at once: in general we should avoid stacking dropdows and buttons vertically, and the places we do need love regardless.

mirka committed:
Increase paddings to 16px

mirka committed:
Fixup changelog

mirka commented:
> It's my sense that we'd want to standardize on 16px padding left and right for all of these 36px items, including Buttons

Gotcha, I've split out the `Button` padding issue to #39431 👍 

As for this PR, I went ahead and changed it to the wider padding (https://github.com/WordPress/gutenberg/pull/39401/commits/126c3d999d184c526efc275d0ceec2aedbc2858b). I'll deal with the right side padding when I fix the chevron problem (#39400).

---

With that, I think all feedback has been accounted for and this PR is ready for final approval 🙏

mirka committed:
Merge branch 'trunk' into enlarge-custom-select-control

ciampo reviewed (approved):
LGTM 🚀 

- Horizontal padding has been adjusted to `16px` in this PR
- Follow-ups:
    - Chevron icon inconsistency tracked separately in #39400
    - Button padding inconsistency tracked separately in #39431
    - Border color inconsistencies tracked separately in #39425
